test/quic_wire_test.c

```diff
@@ -0,0 +1,1224 @@
+{
+    if (!TEST_int_eq(ossl_quic_wire_encode_transport_param_int(pkt, 0x1234, 0x9781), 1))
+        return 0;
+
+    if (!TEST_int_eq(ossl_quic_wire_encode_transport_param_int(pkt, 0x2233, 0x4545), 1))
+        return 0;
+
+    return 1;
+}
+
+static int encode_case_23_dec(PACKET *pkt)
+{
+    uint64_t id = 0, value = 0;
+
+    if (!TEST_int_eq(ossl_quic_wire_decode_transport_param_int(pkt,
+                                                               &id, &value), 1))
+        return 0;
+
+    if (!TEST_uint64_t_eq(id, 0x1234))
+        return 0;
+
+    if (!TEST_uint64_t_eq(value, 0x9781))
+        return 0;
+
+    if (!TEST_int_eq(ossl_quic_wire_decode_transport_param_int(pkt,
+                                                               &id, &value), 1))
+        return 0;
+
+    if (!TEST_uint64_t_eq(id, 0x2233))
+        return 0;
+
+    if (!TEST_uint64_t_eq(value, 0x4545))
+        return 0;
+
+    return 1;
+}
+
+static const uint8_t encode_case_23_expect[] = {
+    0x52, 0x34,
+    0x04,
+    0x80, 0x00, 0x97, 0x81,
+
+    0x62, 0x33,
+    0x04,
+    0x80, 0x00, 0x45, 0x45,
+};
+
+#define ENCODE_CASE(n)                          \
+    {                                           \
+      encode_case_##n##_enc,                    \
+      encode_case_##n##_expect,                 \
+      OSSL_NELEM(encode_case_##n##_expect),     \
+      encode_case_##n##_dec                     \
+    },
+
+static const struct encode_test_case encode_cases[] = {
+    ENCODE_CASE(1)
+    ENCODE_CASE(2)
+    ENCODE_CASE(3)
+    ENCODE_CASE(4)
+    ENCODE_CASE(5)
+    ENCODE_CASE(6)
+    ENCODE_CASE(7)
+    ENCODE_CASE(8)
+    ENCODE_CASE(9)
+    ENCODE_CASE(10)
+    ENCODE_CASE(11)
+    ENCODE_CASE(12)
+    ENCODE_CASE(13)
+    ENCODE_CASE(14)
+    ENCODE_CASE(15)
+    ENCODE_CASE(16)
+    ENCODE_CASE(17)
+    ENCODE_CASE(18)
+    ENCODE_CASE(19)
+    ENCODE_CASE(20)
+    ENCODE_CASE(21)
+    ENCODE_CASE(22)
+    ENCODE_CASE(23)
+};
+
+static int test_wire_encode(int idx)
+{
+    int testresult = 0;
+    WPACKET wpkt;
+    PACKET pkt;
+    BUF_MEM *buf = NULL;
+    size_t written;
+    const struct encode_test_case *c = &encode_cases[idx];
+    int have_wpkt = 0;
+
+    if (!TEST_ptr(buf = BUF_MEM_new()))
+        goto err;
+
+    if (!TEST_int_eq(WPACKET_init(&wpkt, buf), 1))
+        goto err;
+
+    have_wpkt = 1;
+    if (!TEST_int_eq(c->serializer(&wpkt), 1))
+        goto err;
+
+    if (!TEST_int_eq(WPACKET_get_total_written(&wpkt, &written), 1))
+        goto err;
+
+    if (!TEST_mem_eq(buf->data, written, c->expect_buf, c->expect_buf_len))
+        goto err;
+
+    if (!TEST_int_eq(PACKET_buf_init(&pkt, (unsigned char *)buf->data, written), 1))
+        goto err;
+
+    if (!TEST_int_eq(c->deserializer(&pkt), 1))
+        goto err;
+
+    if (!TEST_false(pkt.remaining))
+        goto err;
+
+    testresult = 1;
+err:
+    if (have_wpkt)
+        WPACKET_finish(&wpkt);
+    BUF_MEM_free(buf);
+    return testresult;
+}
+
```

Comment:
We don't seem to have any test cases for decoding illegal frame encodings?

Reply:
Good point. I'll add some.